upgrades/upgradevalidation/validation.go

```diff
@@ -249,6 +249,44 @@ func checkForCharmStoreCharms(_ string, _ StatePool, st State, _ Model) (*Blocke
 	return nil, nil
 }
 
+func checkForCharmsWithNoManifest(_ string, _ StatePool, st State, _ Model) (*Blocker, error) {
+	curls, err := st.AllCharmURLs()
+	if errors.Is(err, errors.NotFound) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	result := set.NewStrings()
+	for _, curl := range curls {
+		if curl == nil {
+			return nil, errors.New("malformed charm in database with no URL")
+		}
+		ch, err := st.Charm(*curl)
+		if err != nil {
+			return nil, errors.Annotatef(err, "getting charm %q", *curl)
```

Comment:
**nitpick:** I tend to avoid multiple dereference of the same pointer, because it tend to explose over the time.

**suggestion:** deref one time, right after the nil check

```go
result := set.NewStrings()
	for _, curlPtr := range curls {
		if curlPtr == nil { return /* ... */	}
		curl := *curlPtr
		// use curl...
```